lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs

```diff
@@ -0,0 +1,184 @@
+        internal void EvaluatorFailed(string evaluatorId)
+        {
+            lock (_lock)
+            {
+                if (!_submittedEvaluators.Contains(evaluatorId))
+                {
+                    Exceptions.Throw(new Exception("Failed evaluator was never submitted"), Logger);
+                }
+
+                _failedEvaluators.Add(evaluatorId);
+                _submittedEvaluators.Remove(evaluatorId);
+            }
+        }
+
+        /// <summary>
+        /// Gives configuration for next evaluator either from failed 
+        /// evaluator or new configuration
+        /// </summary>
+        /// <param name="evaluatorId"></param>
+        /// <returns></returns>
+        internal Tuple<IConfiguration, IConfiguration> GetNextConfiguration(string evaluatorId)
+        {
+            lock (_lock)
+            {
+                if (_submittedEvaluators.Contains(evaluatorId))
+                {
+                    Exceptions.Throw(new Exception("The evaluator is already submitted"), Logger);
+                }
+
+                if (_failedEvaluators.Count == 0 && _assignedPartitionDescriptors >= _numNodes)
+                {
+                    Exceptions.Throw(new Exception("No more configuration can be provided"), Logger);
+                }
+
+                //If some failed id exists return that configuration
+                if (_failedEvaluators.Count != 0)
+                {
+                    string failedEvaluatorId = _failedEvaluators.First();
+                    _failedEvaluators.Remove(failedEvaluatorId);
+                    _configurationProvider[evaluatorId] = _configurationProvider[failedEvaluatorId];
+                }
+                else
+                {
+                    _assignedPartitionDescriptors++;
+
+                    //Checks whether to put update task configuration or map task configuration
+                    if (_assignedPartitionDescriptors == 1)
+                    {
+                        _configurationProvider[evaluatorId] = GetUpdateTaskContextAndServiceConfiguration();
+                    }
+                    else
+                    {
+                        _configurationProvider[evaluatorId] =
+                            GetMapTaskContextAndServiceConfiguration(_partitionDescriptors.Pop());
+                    }
+                }
+
+                _submittedEvaluators.Add(evaluatorId);
+                return _configurationProvider[evaluatorId];
+            }
+        }
+
+        private Tuple<IConfiguration,IConfiguration> GetMapTaskContextAndServiceConfiguration(IPartitionDescriptor partitionDescriptor)
+        {
+            IConfiguration contextConf = _groupCommDriver.GetContextConfiguration();
+            IConfiguration serviceConf = _groupCommDriver.GetServiceConfiguration();
+
+            var codecConfig =
+                TangFactory.GetTang()
+                    .NewConfigurationBuilder(
+                        new[]
```

Comment:
No need for the array. `NewConfigurationBuilder` accepts variable number of arguments.